geomstats/special_orthogonal_group.py

```diff
@@ -557,94 +557,303 @@ def matrix_from_quaternion(self, quaternion):
         assert gs.ndim(rot_mat) == 3
         return rot_mat
 
-    def quaternion_from_yaw_pitch_roll(self, yaw_pitch_roll):
+    def matrix_from_tait_bryan_angles(self, tait_bryan_angles,
+                                      extrinsic_or_intrinsic='extrinsic',
+                                      order='zyx'):
         """
-        Convert a rotation given by the yaw, pitch, roll
-        into a unit quaternion.
+        Convert a rotation given in terms of the tait bryan angles,
+        [angle_x, angle_y, angle_z] in extrinsic (fixed) or
+        intrinsic (moving) coordinate frame into a rotation matrix.
+
+        If the order is zyx, into the rotation matrix rot_mat:
+        rot_mat = X(angle_x).Y(angle_y).Z(angle_z)
+        where:
+        - X(angle_x) is a rotation of angle angle_x around axis x.
+        - Y(angle_y) is a rotation of angle angle_y around axis y.
+        - Z(angle_z) is a rotation of angle angle_z around axis z.
         """
-        assert self.n == 3, ('The quaternion representation'
-                             ' and the yaw-pitch-roll representation'
-                             ' do not exist'
+        assert self.n == 3, ('The tait-bryan angles representation'
+                             ' does not exist'
                              ' for rotations in %d dimensions.' % self.n)
-        yaw_pitch_roll = gs.to_ndarray(yaw_pitch_roll, to_ndim=2)
-        n_yaw_pitch_rolls, _ = yaw_pitch_roll.shape
-
-        yaw = yaw_pitch_roll[:, 0]
-        pitch = yaw_pitch_roll[:, 1]
-        roll = yaw_pitch_roll[:, 2]
-
-        cos_half_yaw = gs.cos(yaw / 2.)
-        sin_half_yaw = gs.sin(yaw / 2.)
-        cos_half_pitch = gs.cos(pitch / 2.)
-        sin_half_pitch = gs.sin(pitch / 2.)
-        cos_half_roll = gs.cos(roll / 2.)
-        sin_half_roll = gs.sin(roll / 2.)
+        tait_bryan_angles = gs.to_ndarray(tait_bryan_angles, to_ndim=2)
+        n_tait_bryan_angles, _ = tait_bryan_angles.shape
+
+        rot_mat = gs.zeros((n_tait_bryan_angles,) + (self.n,) * 2)
+
+        if extrinsic_or_intrinsic == 'extrinsic':
+            if order == 'zyx':
+                angle_x = tait_bryan_angles[:, 0]
+                angle_y = tait_bryan_angles[:, 1]
+                angle_z = tait_bryan_angles[:, 2]
+
+                for i in range(n_tait_bryan_angles):
+                    cos_angle_x = gs.cos(angle_x[i])
+                    sin_angle_x = gs.sin(angle_x[i])
+                    cos_angle_y = gs.cos(angle_y[i])
+                    sin_angle_y = gs.sin(angle_y[i])
+                    cos_angle_z = gs.cos(angle_z[i])
+                    sin_angle_z = gs.sin(angle_z[i])
+
+                    column_1 = [cos_angle_y * cos_angle_z,
```

Comment:
this function is way too big. So many nested if statements and for loops... Should be broken down into pure subfunctions that are easier to unit test too.

Reply:
you mean creating a method: matrix_from_tait_angles_extrinsic, instead of passing extrinsic as parameter?